### PR TITLE
Auto populate DNS names

### PR DESCRIPTION
**Pull request by faisal-memon**
Auto populate DNS names from services attached to pods. Similar to the way its done in the reconcile mode k8s-workload-registrar.
- Creates a field indexer for endpoints, index by pod uid
- When pod comes up we lookup all endpoints with the correct uid
- Adds the following DNS names: `name`, `name.namespace`, `name.namespace.svc`, `name.namespace.svc.clusterdomain`. Not a fan of the bare `name` but added for completeness sake.
- Enabled with `AutoPopulateDNSNames` configurable globally and then individually on each cluster spifffe id.

fixes https://github.com/spiffe/spire-controller-manager/issues/48

### CODE REVIEW

**Review by MarcosDY — COMMENTED** [review]
Thank Faisal!!! 
can you update crd on demo too?

File: migration/README.md [hunk 1]
```diff
@@ -217,23 +217,14 @@ spec:
 
 ### Does SPIRE Controller Manager automatically populate DNS Names of Services a Pod is attached to?
 
-SPIRE Controller Manager doesn't monitor Endpoints like Kubernetes Workload Registrar did, so it won't do this automatically. A workaround is to use the `app` label to populate DNS Names using `dnsNameTemplates` field of the [ClusterSPIFFEID CRD](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md), assuming you are using `app` as your selector and it matches the name of the `Service`.
+Yes, this is enabled with the sample configuration in this migration guide. To enable in your deployment follow these steps:
 
-```yaml
```
[the remark this answers is not in the document]
Comment (faisal-memon) [reply migration/README.md:222]: Yes dnsNameTemplate still usable if you want custom dns names. This example is just a workaround for this missing feature. The example right above talks about how to use dns name templates more generally.

File: controllers/endpoints_controller.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2021 SPIRE Authors.
```
[the remark this answers is not in the document]
Comment (faisal-memon) [reply controllers/endpoints_controller.go]: updated to 2023

File: migration/README.md [hunk 3]
```diff
@@ -217,23 +217,14 @@ spec:
...
-      spiffe.io/spiffe-id: "true"
-  dnsNameTemplates: ["{{ index .PodMeta.Labels \"app\" }}.{{ .PodMeta.Namespace }}.svc.cluster.local"]
+1. Enable `autoPopulateDNSNames` globally in the [Controller Manager Config](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md). See [example](config/spire-controller-manager-config.yaml).
+1. For each  [ClusterSPIFFEID](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md) you to auto populate DNS names for, set the `autoPopulateDNSNames` field there. See [example](config/clusterspiffeid.yaml).
 
-```
 
-If you require these DNS Names to be automatically populated, please update [#48](https://github.com/spiffe/spire-controller-manager/issues/48) with your use case.
+> **Note**
+> Spire Controller Manager 0.2.3 or later is required to auto populate DNS names.
```
[the remark this answers is not in the document]
Comment (faisal-memon) [reply migration/README.md]: done

File: controllers/pod_controller.go [hunk 4]
```diff
@@ -55,6 +57,32 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *PodReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Index endpoints by UID. Later when we reconcile the Pod this will make it easy to find the associated endpoints
+	// and auto populate DNS names.
+	if r.AutoPopulateDNSNames {
+		err := mgr.GetFieldIndexer().IndexField(context.Background(), &corev1.Endpoints{}, reconciler.EndpointUID, func(rawObj client.Object) []string {
+			endpoints := rawObj.(*corev1.Endpoints)
```
[the remark this answers is not in the document]
Comment (faisal-memon) [reply controllers/pod_controller.go]: done

File: controllers/endpoints_controller.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	IgnoreNamespaces stringset.StringSet
+}
+
+//+kubebuilder:rbac:groups=spire.spiffe.io,resources=clusterspiffeids,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=spire.spiffe.io,resources=clusterspiffeids/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=spire.spiffe.io,resources=clusterspiffeids/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch
+//+kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch
+//+kubebuilder:rbac:groups="",resources=endpoints,verbs=get;list;watch
+//+kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch
```
[the remark this answers is not in the document]
Comment (faisal-memon) [reply controllers/endpoints_controller.go]: done

File: pkg/spireentry/entries.go [hunk 6]
```diff
@@ -58,18 +59,12 @@ func renderPodEntry(spec *spirev1alpha1.ParsedClusterSPIFFEIDSpec, node *corev1.
...
-		if _, exists := dnsNamesSet[dnsName]; !exists {
-			dnsNamesSet[dnsName] = struct{}{}
-			dnsNames = append(dnsNames, dnsName)
-		}
+	dnsNames = addServiceDNSNames(dnsNames, dnsNamesSet, endpointsList, clusterDomain)
+	dnsNames, err = addRenderedDNSNames(dnsNames, dnsNamesSet, spec.DNSNameTemplates, data)
+	if err != nil {
+		return nil, err
 	}
+	sort.Strings(dnsNames) // Sort the list to provide consistent results
```
[the remark this answers is not in the document]
Comment (faisal-memon) [reply pkg/spireentry/entries.go]: Thanks @MarcosDY for catching this. Its updated to only sort the auto added dns names. The templated ones will be first and in the order specified in the crd.

File: pkg/spireentry/entries_test.go [hunk 7]
```diff
@@ -59,8 +75,23 @@ func TestRenderPodEntry(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, entry.ParentID.String(), parentID.String())
 
-	// DNS names rendered correctly and are unique
-	require.Len(t, entry.DNSNames, len(spec.DNSNameTemplates)-1)
+	// DNS names are unique
+	dnsNamesSet := make(map[string]struct{})
+	for _, dnsName := range entry.DNSNames {
```
[the remark this answers is not in the document]
Comment (faisal-memon) [reply pkg/spireentry/entries_test.go:80]: Added a length check to the test.

**Comment by faisal-memon** [comment]
All comments have been addressed. The demo crds have been updated.

File: api/v1alpha1/controllermanagerconfig_types.go [hunk 8]
```diff
@@ -32,6 +32,9 @@ type ControllerManagerConfig struct {
 	// ControllerManagerConfigurationSpec returns the contfigurations for controllers
 	ControllerManagerConfigurationSpec `json:",inline"`
 
+	// AutoPopulateDNSNames indicates whether or not to auto populate service DNS names.
+	AutoPopulateDNSNames bool `json:"autoPopulateDNSNames,omitempty"`
+
```
Comment (faisal-memon) [inline api/v1alpha1/controllermanagerconfig_types.go:37]: I think this extra controller manager configurable is redundant. The feature can be toggled on the crd.

Comment (faisal-memon) [reply api/v1alpha1/controllermanagerconfig_types.go]: Removed the extra configurable.

File: controllers/pod_controller.go [hunk 9]
```diff
@@ -61,6 +61,36 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *PodReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Index endpoints by UID. Later when we reconcile the Pod this will make it easy to find the associated endpoints
+	// and auto populate DNS names.
+	ctx := context.Background()
+	err := mgr.GetFieldIndexer().IndexField(ctx, &corev1.Endpoints{}, reconciler.EndpointUID, func(rawObj client.Object) []string {
```
Comment (faisal-memon) [inline controllers/pod_controller.go:67]: Not sure if context.Background() is correct here or if we should use something else.

File: pkg/spireentry/entries.go [hunk 10]
```diff
@@ -160,6 +166,23 @@ func renderDNSName(tmpl *template.Template, data *templateData) (string, error)
 	return rendered, nil
 }
 
+func addServiceDNSNames(dnsNames []string, dnsNamesSet map[string]struct{}, endpointsList *corev1.EndpointsList, clusterDomain string) []string {
```
Comment (MarcosDY) [inline pkg/spireentry/entries.go:169]: what do you think about this function returning the list of dns from endpoints (only) and sorted.
and in the caller add the new dns to the list?
```
dsnNames = append(dnsNames, dnsNamesFromEndpoints( dnsNamesSet, endpointsList, clusterDomain))
``` 

that wil simplify this code

File: migration/README.md [hunk 11]
```diff
@@ -217,23 +218,14 @@ spec:
...
-metadata:
-  name: federation
-spec:
-  spiffeIDTemplate: "spiffe://{{ .TrustDomain }}/ns/{{ .PodMeta.Namespace }}/sa/{{ .PodSpec.ServiceAccountName }}"
-  podSelector:
-    matchLabels:
-      spiffe.io/spiffe-id: "true"
-  dnsNameTemplates: ["{{ index .PodMeta.Labels \"app\" }}.{{ .PodMeta.Namespace }}.svc.cluster.local"]
+1. Enable `autoPopulateDNSNames` globally in the [Controller Manager Config](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md). See [example](config/spire-controller-manager-config.yaml).
+1. For each  [ClusterSPIFFEID](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md) you to auto populate DNS names for, set the `autoPopulateDNSNames` field there. See [example](config/clusterspiffeid.yaml).
```
[the remark this answers is not in the document]
Comment (faisal-memon) [reply migration/README.md]: Fixed

[the remark this answers is not in the document]
Comment (faisal-memon) [reply migration/README.md]: Have each listed as 1. i think is better. Then you don't have to update every line if you add an item in the middle.

File: pkg/spireentry/entries.go (hunk 10 above)
Comment (faisal-memon) [reply pkg/spireentry/entries.go]: Change made.

File: controllers/pod_controller.go [hunk 12]
```diff
@@ -61,6 +61,36 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl
...
+		log := log.FromContext(ctx)
+		endpoints, ok := rawObj.(*corev1.Endpoints)
+		if !ok {
+			log.Error(nil, "unexpected type indexing fields", "type", fmt.Sprintf("%T", rawObj), "expecteed", "*corev1.Endpoints")
```
[the remark this answers is not in the document]
Comment (faisal-memon) [reply controllers/pod_controller.go:71]: should not happen ever

File: migration/README.md [hunk 13]
```diff
@@ -217,23 +218,14 @@ spec:
...
-kind: ClusterSPIFFEID
-metadata:
-  name: federation
-spec:
-  spiffeIDTemplate: "spiffe://{{ .TrustDomain }}/ns/{{ .PodMeta.Namespace }}/sa/{{ .PodSpec.ServiceAccountName }}"
-  podSelector:
-    matchLabels:
-      spiffe.io/spiffe-id: "true"
-  dnsNameTemplates: ["{{ index .PodMeta.Labels \"app\" }}.{{ .PodMeta.Namespace }}.svc.cluster.local"]
+1. Enable `autoPopulateDNSNames` globally in the [Controller Manager Config](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md). See [example](config/spire-controller-manager-config.yaml).
```
[the remark this answers is not in the document]
Comment (faisal-memon) [reply migration/README.md]: done

File: controllers/pod_controller.go (hunk 9 above)
Comment (faisal-memon) [reply controllers/pod_controller.go]: Thanks, I missed that one. Change is in.

**Comment by faisal-memon** [comment]
Docs updated. Trying to rebase to fix the dco causes a lot merge conflicts. Whenever this is approved Ill squash and rebase and then fix the dco.

File: controllers/pod_controller.go [hunk 14]
```diff
@@ -60,7 +60,36 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl
 }
 
 // SetupWithManager sets up the controller with the Manager.
-func (r *PodReconciler) SetupWithManager(mgr ctrl.Manager) error {
+func (r *PodReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	// Index endpoints by UID. Later when we reconcile the Pod this will make it easy to find the associated endpoints
+	// and auto populate DNS names.
+	err := mgr.GetFieldIndexer().IndexField(ctx, &corev1.Endpoints{}, reconciler.EndpointUID, func(rawObj client.Object) []string {
+		log := log.FromContext(ctx)
```
Comment (MarcosDY) [inline controllers/pod_controller.go:67]: log is only used whn error, can you move into if?

**Review by MarcosDY — COMMENTED** [review]
looks good, just a simple nit, and can you solve DCO?

File: controllers/pod_controller.go (hunk 14 above)
Comment (faisal-memon) [reply controllers/pod_controller.go]: Done

**Comment by faisal-memon** [comment]
@MarcosDY all comments addressed, dco fixed.

**Review by MarcosDY — APPROVED** [review]

**Pull request merged**